### PR TITLE
Add options to fix clipping and apply peak normalization to the generated soundscape

### PR DESCRIPTION
**Pull request by justinsalamon**
Right now, it's very easy for the user to generate a soundscape the clips if they're not careful with their choice of `ref_db` and the SNR values they give to foreground events.

This PR adds two new optional parameters to both `generate` and `generate_from_jams`:
* `fix_clipping` - if True (default=False), will apply peak normalization in the case that the soundscape is clipping
* `peak_normalization` - allows the user to apply peak normalization (abs(max) sample = 1) even if the soundscape isn't clipping

This PR addresses #131 

Checklist:

generate:
- [x] modify soundscape audio
- [x] modify event audio
- [x] jams: update ref_db
- [x] jams: store fix_clipping and peak_normalization bools
- [x] jams: store scale_factor?

generate_from_jams:
- [x] modify soundscape audio
- [x] modify event audio
- [x] jams: update ref_db
- [x] jams: store fix_clipping and peak_normalization bools
- [x] jams: store scale_factor?

testing:
- [x] peak_normalize
- [x] generate
- [x] generate_from_jams

### CODE REVIEW

**Comment by justinsalamon** [comment]
NOTE: missing unite tests, coverage at 98% right now.

File: scaper/audio.py [hunk 1]
```diff
@@ -1,18 +1,15 @@
 # CREATED: 4/23/17 15:37 by Justin Salamon <justin.salamon@nyu.edu>
 
-'''
+"""
 Utility functions for audio processing using FFMPEG (beyond sox). Based on:
```
Comment (pseeth) [inline scaper/audio.py:4]: This is my bad, but we can update this top comment since we don't use ffmpeg anymore.

File: scaper/core.py [hunk 2]
```diff
@@ -214,7 +215,7 @@ def generate_from_jams(jams_infile,
 
     # Cast ann.sandbox.scaper to a Sandbox object
     ann.sandbox.scaper = jams.Sandbox(**ann.sandbox.scaper)
-    soundscape_audio, event_audio_list = \
+    soundscape_audio, event_audio_list, scale_factor = \
```
Comment (pseeth) [inline scaper/core.py:218]: This returns `scale_factor`, but we're not saving the `fix_clipping` and `peak_normalize` vals anywhere right? The logic being that `ref_db` was updated when this JAM was created, I think..

Hmm, what if someone does `sc.generate` with `no_audio=True`, but with `peak_normalize=True` or `fix_clipping=True`. We may have to store it in the JAMS after all.

File: scaper/core.py [hunk 3]
```diff
@@ -1913,29 +1939,86 @@ def _generate_audio(self, audio_path, ann, reverb=None,
...
+                    warnings.warn('Soundscape audio is clipping!',
+                                  ScaperWarning)
+
+                if peak_normalization or (clipping and fix_clipping):
+
+                    # normalize soundscape audio and scale event audio
+                    soundscape_audio, event_audio_list, scale_factor = \
+                        peak_normalize(soundscape_audio, event_audio_list)
+
+                    warnings.warn(
```
Comment (pseeth) [inline scaper/core.py:1969]: This would warn every time if `peak_normalization=True`, which seems excessive.

File: scaper/core.py [hunk 4]
```diff
@@ -1913,29 +1939,86 @@ def _generate_audio(self, audio_path, ann, reverb=None,
...
+                        'Peak normalization applied (scale factor = {})'.format(
```
Comment (pseeth) [inline scaper/core.py:1970]: Can we also report the change that happened to `ref_db`? It would be `20*np.log10(scale_factor)`.

File: scaper/core.py [hunk 5]
```diff
@@ -1768,6 +1793,7 @@ def _generate_audio(self, audio_path, ann, reverb=None,
         # List for storing all generated audio (one array for every event)
         soundscape_audio = None
         event_audio_list = []
+        scale_factor = None
```
Comment (pseeth) [inline scaper/core.py:1796]: Instead of `None`, you could say the `scale_factor` defaults to `1.0`, which might make the logic in the JAMS annotation easier to handle.

**Review by pseeth — CHANGES_REQUESTED** [review]
Looking good so far, I've left some comments and thoughts.

File: scaper/audio.py (hunk 1 above)
Comment (justinsalamon) [reply scaper/audio.py:4]: cool, removed

File: scaper/core.py (hunk 3 above)
Comment (justinsalamon) [reply scaper/core.py:1969]: fair point, I'll change it so it only throws when it's fixing clipping explicitly

Comment (justinsalamon) [reply scaper/core.py:1969]: changed

File: scaper/core.py [hunk 6]
```diff
@@ -1966,16 +1964,23 @@ def _generate_audio(self,
                     soundscape_audio, event_audio_list, scale_factor = \
                         peak_normalize(soundscape_audio, event_audio_list)
 
-                    warnings.warn(
-                        'Peak normalization applied (scale factor = {})'.format(
-                            scale_factor),
-                        ScaperWarning)
+                    ref_db_drop = 20 * np.log10(scale_factor)
```
Comment (pseeth) [inline scaper/core.py:1967]: I'd say `ref_db_change` instead of `ref_db_drop`, as if `scale_factor > 1.0`, then it's actually a gain instead of a drop.

File: scaper/core.py [hunk 7]
```diff
@@ -1966,16 +1964,23 @@ def _generate_audio(self,
...
+
+                    if clipping and fix_clipping:
+                        warnings.warn(
+                            'Peak normalization applied to fix clipping with '
+                            'scale factor = {}. The actual ref_db of the '
+                            'generated soundscape audio will be lower by '
+                            'approximately {:.2f}dB with respect to the target '
+                            'ref_db of {})'.format(
+                                scale_factor, ref_db_drop, self.ref_db),
```
Comment (pseeth) [inline scaper/core.py:1976]: Let's say "change by approximately " to account for positive or negative db changes.

File: tests/test_audio.py [hunk 8]
```diff
@@ -95,3 +100,73 @@ def test_match_sample_length():
...
+                    print("actual scale factor: {}".format(scale_factor))
+
+                    # test scale factor
+                    max_sample_audio = np.max(np.abs(audio_norm))
+                    print("max sample audio norm: {}".format(max_sample_audio))
+                    assert np.allclose(scale_factor, estimated_scale_factor,
+                                       atol=1e-03)
+
+                    # test soundscape audio
+                    assert max_sample_audio < 1.0
```
Comment (pseeth) [inline tests/test_audio.py:156]: `<=`?

File: tests/test_audio.py [hunk 9]
```diff
@@ -95,3 +100,73 @@ def test_match_sample_length():
...
+                    print("allclose on factors: {}".format(np.allclose(
+                        scale_factor, estimated_scale_factor)))
+
+                    print("actual scale factor: {}".format(scale_factor))
+
+                    # test scale factor
+                    max_sample_audio = np.max(np.abs(audio_norm))
+                    print("max sample audio norm: {}".format(max_sample_audio))
+                    assert np.allclose(scale_factor, estimated_scale_factor,
+                                       atol=1e-03)
```
Comment (pseeth) [inline tests/test_audio.py:153]: `atol` seems high to me?

**Comment by pseeth** [comment]
If we can unpin pyrsistent in this PR as well, that'd be great. We pinned it to be compat with 2.7 and 3.4, but we dropped support for those a few versions ago, so we can unpin.

File: tests/test_core.py [hunk 10]
```diff
@@ -2046,3 +2085,92 @@ def test_scaper_with_short_background():
...
+            event_time=('const', 0),
+            event_duration=('const', 10),
+            snr=('const', 20),
+            pitch_shift=None,
+            time_stretch=None)
+
+        # extreme clipping
+        sc_extreme = scaper.Scaper(10, FG_PATH, BG_PATH, random_state=0)
+        sc_extreme.sr = 16000
+        sc_extreme.ref_db = -20
```
Comment (pseeth) [inline tests/test_core.py:2110]: It might be stronger to iterate over a few `ref_db` values instead of just one?

File: docs/changes.rst [hunk 11]
```diff
@@ -2,6 +2,14 @@
 
 Changelog
 ---------
+v1.6.3
+~~~~~~
+- Scaper.generate now accepts two new optional arguments for controlling audio clipping and normalization:
+    - fix_clipping: if True and the soundscape audio is clipping, it will be peak normalized and all isolated events will be scaled accordingly.
+    - peak_normalization: if True, sounscape audio will be peak normalized regardless of whether it's clipping or not and all isolated events will be scaled accordingly.
+- All generate arguments are now documented in the scaper sandbox inside the JAMS annotation.
+- Furthermore, we also document in the JAMS: the scale factor used for peak normalization, the change in ref_db, and the actual ref_db of the generated audio.
```
Comment (pseeth) [inline docs/changes.rst:11]: Should make sure this ends up looking okay in ReadTheDocs.

File: scaper/audio.py [hunk 12]
```diff
@@ -104,5 +96,42 @@ def match_sample_length(audio_path, duration_in_samples):
...
+    scaled_soundscape_audio : np.ndarray
+        The peak normalized soundscape audio.
+    scaled_event_audio_list : list
+        List of np.ndarrays containing the scaled audio samples of
+        each isolated foreground event. All events are scaled by scale_factor.
+    scale_factor : float
+        The scale factor used to peak normalize the soundscape audio.
+    """
+    eps = 1e-10
+    max_sample = np.max(np.abs(soundscape_audio))
```
Comment (pseeth) [inline scaper/audio.py:127]: Can we add an edge case in the tests where all the source files we're putting into the soundscape are silent? Just want to make sure nothing breaks. The scaling factor will end up at `1e10` in that case, but it should be fine when multiplied by all zeros. But might be a good code path to make sure this regularization is working going into the future.

File: scaper/core.py [hunk 13]
```diff
@@ -2103,12 +2221,38 @@ def generate(self,
...
                                      save_isolated_events=save_isolated_events,
                                      isolated_events_path=isolated_events_path,
-                                     disable_sox_warnings=disable_sox_warnings)
+                                     disable_sox_warnings=disable_sox_warnings,
+                                     fix_clipping=fix_clipping,
+                                     peak_normalization=peak_normalization)
+
+        # TODO: Stick to heavy handed overwriting for now, in the future we
+        #  should consolidate this with what happens inside _instantiate().
+        # print("THIS!!!", allow_repeated_label, type(allow_repeated_label))
```
[the remark this answers is not in the document]
Comment (pseeth) [reply scaper/core.py:2238]: The print comment, that is.

File: scaper/core.py (hunk 2 above)
Comment (justinsalamon) [reply scaper/core.py]: we're now storing these in the jams

File: scaper/core.py (hunk 4 above)
Comment (justinsalamon) [reply scaper/core.py]: done

File: scaper/core.py (hunk 5 above)
Comment (justinsalamon) [reply scaper/core.py]: done

File: scaper/core.py (hunk 6 above)
Comment (justinsalamon) [reply scaper/core.py]: haha didn't see this comment till now but I had the very same thought and changed it!

File: scaper/core.py (hunk 7 above)
Comment (justinsalamon) [reply scaper/core.py]: again, snap!

File: tests/test_audio.py (hunk 8 above)
Comment (justinsalamon) [reply tests/test_audio.py:156]: since we add an eps, it should never actually reach 1.0

File: tests/test_audio.py (hunk 9 above)
Comment (justinsalamon) [reply tests/test_audio.py:153]: anything larger and it fails due to floating point inaccuracies, I checked.

Comment (justinsalamon) [reply tests/test_audio.py:153]: But feel free to test on your end too to confirm

File: scaper/core.py [hunk 14]
```diff
@@ -209,13 +210,21 @@ def generate_from_jams(jams_infile,
     sc.fade_in_len = ann.sandbox.scaper['fade_in_len']
     sc.fade_out_len = ann.sandbox.scaper['fade_out_len']
 
-    # Generate audio and save to disk
+    # Pull generation parameters from annotation
     reverb = ann.sandbox.scaper['reverb']
+    fix_clipping = ann.sandbox.scaper['fix_clipping']
+    peak_normalization = ann.sandbox.scaper['peak_normalization']
```
[the remark this answers is not in the document]
Comment (justinsalamon) [reply scaper/core.py:216]: NEw code:
```python
    if 'fix_clipping' in ann.sandbox.scaper.kesy():
        fix_clipping = ann.sandbox.scaper['fix_clipping']
    else:
        fix_clipping = False

    if 'peak_normalization' in ann.sandbox.scaper.kesy():
        peak_normalization = ann.sandbox.scaper['peak_normalization']
    else:
        peak_normalization = False
```

[the remark this answers is not in the document]
Comment (justinsalamon) [reply scaper/core.py:216]: fixed typo `keys`

File: scaper/core.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (justinsalamon) [reply scaper/core.py:2238]: removed the print. the TODO comment should stick around I believe

File: tests/data/regression/soundscape_20200501_44100.jams [hunk 15]
```diff
@@ -235,9 +235,23 @@
...
+          "save_isolated_events": false,
+          "isolated_events_path": null,
+          "disable_sox_warnings": true,
+          "no_audio": false,
+          "txt_path": "tests/data/regression/soundscape_20200501_44100.txt",
+          "txt_sep": "\t",
+          "disable_instantiation_warnings": true,
+          "peak_normalization_scale_factor": 1.0,
+          "ref_db_change": 0,
+          "ref_db_generated": -50
```
[the remark this answers is not in the document]
Comment (justinsalamon) [reply tests/data/regression/soundscape_20200501_44100.jams:254]: yes we should, but... can we open an issue for this and do it AFTER the tutorial? 😬

File: tests/test_audio.py [hunk 16]
```diff
@@ -95,3 +100,73 @@ def test_match_sample_length():
...
+                        max_sample_event = np.max(np.abs(event_audio))
+
+                        if not np.allclose(
+                                max_sample_event, A * factor * scale_factor):
+                            print(max_sample_audio, max_sample_event,
+                                  A * factor * scale_factor)
+
+                        assert np.allclose(max_sample_event,
+                                           A * factor * scale_factor,
+                                           atol=1e-3)
```
[the remark this answers is not in the document]
Comment (justinsalamon) [reply tests/test_audio.py:172]: for some reason GH UI isn't letting me respond to the previous 2 comments, so answering here for all 3:
* low atol because any higher and test fails, I think it's floating point issues
* < 1.0 because we add eps so it should never hit 1

File: tests/test_core.py [hunk 17]
```diff
@@ -2046,3 +2085,92 @@ def test_scaper_with_short_background():
         audio1, sr = soundfile.read(wav1_file.name)
         audio2, sr = soundfile.read(wav2_file.name)
         assert np.allclose(audio1, audio2)
+
+
+def test_clipping_and_normalization():
```
[the remark this answers is not in the document]
Comment (justinsalamon) [reply tests/test_core.py:2090]: I need to add a no_audio case for generate_from_jams, but in the long run we should add no_audio cases elsewhere. Shall we open an issue?

File: tests/test_core.py (hunk 10 above)
Comment (justinsalamon) [reply tests/test_core.py:2110]: Well it only clips if SNR is high enough with respect to red_db, meaning I'd have to also change the SNR value, meaning I'm not sure that would actually test anything different?

File: docs/changes.rst (hunk 11 above)
Comment (justinsalamon) [reply docs/changes.rst:11]: yeah this was a bit of a gamble.. if it looks bad we can open an issue to fix after this PR?

File: scaper/audio.py (hunk 12 above)
Comment (justinsalamon) [reply scaper/audio.py:127]: Good point, but we'd have to create more regression audio data for that, so let's open an issue for this and leave it for a future PR to strengthen testing

File: scaper/core.py [hunk 18]
```diff
@@ -209,16 +210,45 @@ def generate_from_jams(jams_infile,
...
+    # should consolidate this with what happens inside _instantiate().
+    ann.sandbox.scaper.reverb = reverb
+    ann.sandbox.scaper.fix_clipping = fix_clipping
+    ann.sandbox.scaper.peak_normalization = peak_normalization
+    ann.sandbox.scaper.save_isolated_events = save_isolated_events
+    ann.sandbox.scaper.isolated_events_path = isolated_events_path
+    ann.sandbox.scaper.disable_sox_warnings = disable_sox_warnings
+    ann.sandbox.scaper.peak_normalization_scale_factor = scale_factor
+    ann.sandbox.scaper.ref_db_change = ref_db_change
+    ann.sandbox.scaper.ref_db_generated = sc.ref_db + ref_db_change
```
Comment (pseeth) [inline scaper/core.py:250]: This code is duplicated between here and in `generate`. Might be worth refactoring at some point.

File: scaper/core.py [hunk 19]
```diff
@@ -1913,29 +1986,94 @@ def _generate_audio(self, audio_path, ann, reverb=None,
...
+                            'audio may not perfectly match their specified values.',
+                            ScaperWarning
+                        )
+
+                # Apply effects and reshape
+                soundscape_audio = tfm.build_array(
+                    input_array=soundscape_audio,
+                    sample_rate_in=self.sr,
+                )
+                soundscape_audio = soundscape_audio.reshape(-1, self.n_channels)
```
Comment (pseeth) [inline scaper/core.py:2041]: So...just noticed that we do all of the clipping and peak normalization stuff BEFORE we apply reverb to the soundscape. Should these operations be reversed? Reverb could cause things to clip, or it could cause it to not be peak normalized anymore, etc. Sorry, I just noticed this was happening.

File: scaper/core.py [hunk 20]
```diff
@@ -1913,29 +1986,94 @@ def _generate_audio(self, audio_path, ann, reverb=None,
...
-                        '{:s}{:d}_{:s}{:s}'.format(
-                            e.value['role'], _role_count, e.value['label'], ext))
-                    role_counter[e.value['role']] += 1
-                    
                     if not os.path.exists(event_folder):
-                        # In Python 3.2 and above we could do 
+                        # In Python 3.2 and above we could do
                         # os.makedirs(..., exist_ok=True) but we test back to
                         # Python 2.7.
                         os.makedirs(event_folder)
```
Comment (pseeth) [inline scaper/core.py:2060]: The comment here suggests that we could get rid of the `if` block, as we now only support Python 3.5+.

File: scaper/core.py [hunk 21]
```diff
@@ -1944,41 +2082,24 @@ def _generate_audio(self, audio_path, ann, reverb=None,
...
-                    sample_rate_in=self.sr,
-                )
-                soundscape_audio = soundscape_audio.reshape(-1, self.n_channels)
-
-                # Save to disk if output path provided
-                if audio_path is not None:
-                    soundfile.write(audio_path, soundscape_audio, self.sr, subtype='PCM_32')
-
         # Document output paths
+        # TODO: this is redundant with data stored in ann.sandbox.scaper.generate,
```
Comment (pseeth) [inline scaper/core.py:2086]: Don't think this matters too much but the comment says things are stored in `ann.sandbox.scaper.generate` when I think we store everything in `ann.sandbox.scaper`?

File: scaper/core.py [hunk 22]
```diff
@@ -2103,13 +2241,38 @@ def generate(self,
...
+        ann.sandbox.scaper.save_isolated_events = save_isolated_events
+        ann.sandbox.scaper.isolated_events_path = isolated_events_path
+        ann.sandbox.scaper.disable_sox_warnings = disable_sox_warnings
+        ann.sandbox.scaper.no_audio = no_audio
+        ann.sandbox.scaper.txt_path = txt_path
+        ann.sandbox.scaper.txt_sep = txt_sep
+        ann.sandbox.scaper.disable_instantiation_warnings = disable_instantiation_warnings
+        ann.sandbox.scaper.peak_normalization_scale_factor = scale_factor
+        ann.sandbox.scaper.ref_db_change = ref_db_change
+        ann.sandbox.scaper.ref_db_generated = self.ref_db + ref_db_change
```
Comment (pseeth) [inline scaper/core.py:2274]: Duplicates code from `generate_from_jams` (this code is a superset of that code). So might be worth refactoring, as I mentioned before.

**Review by pseeth — CHANGES_REQUESTED** [review]
Looks pretty good, there are a few tiny things that should be looked at before we merge, I think. See comments.

File: scaper/core.py (hunk 18 above)
Comment (justinsalamon) [reply scaper/core.py:250]: Agreed - I'll create an issue for it so we address in a future PR

Comment (justinsalamon) [reply scaper/core.py:250]: https://github.com/justinsalamon/scaper/issues/135

File: scaper/core.py (hunk 19 above)
Comment (justinsalamon) [reply scaper/core.py:2041]: I've re-ordered the operations. Current tests are passing. @pseeth is checking now whether applying reverb to a clipping signal and then normalization works as expected. Assuming it does, we'll stick to this ordering. We need to add unit tests to explicitly test for this scenario, I've added a comment on #134

Comment (pseeth) [reply scaper/core.py:2041]: After looking into this, we shouldn't normalize. See #136.

File: scaper/core.py (hunk 20 above)
Comment (justinsalamon) [reply scaper/core.py:2060]: updated

File: scaper/core.py (hunk 21 above)
Comment (justinsalamon) [reply scaper/core.py:2086]: we're not storing anything in `scaper.generate` (that was an earlier idea), but the duplicity still exists (for compatibility, for now). I've updated the comment to reflect this.

File: scaper/core.py (hunk 19 above)
Comment (justinsalamon) [reply scaper/core.py:2041]: @pseeth has checked and the re-ordering fails. I've reverted to the original order: sum --> normalize --> reverb.

File: scaper/core.py (hunk 22 above)
Comment (justinsalamon) [reply scaper/core.py:2274]: I've created #135 so we address this in a future PR.

**Comment by justinsalamon** [comment]
ok @pseeth I believe I've addressed all of your comments. Let's CR and SqM.

**Review by pseeth — APPROVED** [review]
